sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala

```diff
@@ -0,0 +1,456 @@
+      if (summary.count > 0) new GenericArrayData(percentiles.map { summary.query(_) }) else null
+    }
+    else {
+      // return the result as a double, or return null for empty inputs
+      if (summary.count > 0) summary.query(percentiles.head) else null
+    }
+  }
+
+  private def childrenSQL: String =
+    if (bExpr.isDefined) {
+      s"${child.sql}, ${percentilesExpr.toString}, ${bExpr.get.toString}"
+    }
+    else {
+      s"${child.sql}, ${percentilesExpr.toString}"
+    }
+
+  override def sql: String = s"$prettyName($childrenSQL)"
+
+  override def sql(isDistinct: Boolean): String = {
+    val distinct = if (isDistinct) "DISTINCT " else ""
+    s"$prettyName($distinct$childrenSQL)"
+  }
+}
+
+object PercentileApprox {
+
+  // Using 1000 as a default achieves a relativeError of 0.001
+  private[sql] val B_DEFAULT = 1000
+
+  // Our `PercentileApprox` function takes a `B` parameter, but the underlying GK algorithm takes
+  // a `relativeError` parameter, so we need to convert `B` to `relativeError`.
+  // Please refer to SPARK-16283 for details.
+  private[sql] def bToRelativeError(B: Int): Double = Math.max(1.0d / B, 0.001)
+
+  /**
+   * Validates the percentile(s) expression and extract the percentile(s).
+   * Returns the extracted percentile(s) and an indicator of whether it's an array.
+   */
+  private def validatePercentilesLiteral(exp: Expression): (Seq[Double], Boolean) = {
+    def withinRange(v: Double): Boolean = 0.0 <= v && v <= 1.0
+    exp match {
+      case Literal(f: Float, FloatType) if withinRange(f) => (Seq(f.toDouble), false)
+      case Literal(d: Double, DoubleType) if withinRange(d) => (Seq(d), false)
+      case Literal(dec: Decimal, _) if withinRange(dec.toDouble) => (Seq(dec.toDouble), false)
+
+      case CreateArray(children: Seq[Expression]) if (children.length > 0) =>
+        (children.map(_ match {
+          case Literal(f: Float, FloatType) if withinRange(f) => f.toDouble
+          case Literal(d: Double, DoubleType) if withinRange(d) => d
+          case Literal(dec: Decimal, _) if withinRange(dec.toDouble) => dec.toDouble
+          case _ =>
+            throw new AnalysisException(
+              "The second argument should be a double literal or an array of doubles, and should " +
+                "be within range [0.0, 1.0]")
+        }), true)
+
+      case _ =>
+        throw new AnalysisException(
+          "The second argument should be a double literal or an array of doubles, and should " +
+            "be within range [0.0, 1.0]")
+    }
+  }
+
+  /** Validates the B expression and extract its value. */
+  private def validateBLiteral(exp: Expression): Int = exp match {
+    case Literal(i: Int, IntegerType) if i > 0 => i
+
+    case _ =>
+      throw new AnalysisException("The third argument should be a positive integer literal")
+  }
+}
+
+/**
+ * Helper class to compute approximate quantile summary.
+ * This implementation is based on the algorithm proposed in the paper:
```

Comment:
thanks!

module `sql/core` depends on module `sql/catalyst`, but module `sql/catalyst` does not depend on module `sql/core`; so in order to enable expressions like `PercentileApprox` (which lives in `sql/catalyst`) to use of the G-K algorithm, the algorithm should probably live in `sql/catalyst`. As a reference, algorithms like HyperLogLog also live in `sql/catalyst`(please refer to [HyperLogLogPlusPlus](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala)).

@thunterdb does this make sense?